cedar-lean/Cedar/Thm/Lemmas/Typechecker.lean

```diff
@@ -0,0 +1,167 @@
+/-
+ Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Spec
+import Cedar.Validation
+
+/-!
+This file contains useful definitions and lemmas about the `Typechecker` functions.
+
+todo: fill in `sorry`s. Some invariants may need to be adjusted. The current
+definitions are an informed guess based on the corresponding Dafny proof.
+-/
+
+namespace Cedar.Thm
+
+open Cedar.Data
+open Cedar.Spec
+open Cedar.Validation
+
+def InstanceOfBoolType (b : Bool) (bty : BoolType) : Prop :=
+  match (b, bty) with
+  | (true, .tt) => True
+  | (false, .ff) => True
+  | (_, .anyBool) => True
+  | _ => False
+
+def InstanceOfEntityType (e : EntityUID) (ety: EntityType) : Prop :=
+  ety = e.ty
+
+def InstanceOfExtType (e : Ext) (ety : ExtType) : Prop :=
+  match (e, ety) with
+  | (.decimal _, .decimal) => True
+  | (.ipaddr _, .ipAddr) => True
+  | _ => False
+
+inductive InstanceOfType : Value → CedarType → Prop :=
+  | instance_of_bool (b : Bool) (bty : BoolType) : InstanceOfBoolType b bty →
```

Comment:
Do we want to name the hypotheses in the inductive predicate definitions (here and below)?  

If we do, Lean will automatically use those names in proofs, which is a nice bonus.

Something like:

```
| instance_of_bool (b: Bool) (bty : BoolType)  
   (h₁ : InstanceOfBoolType b bty) : 
   InstanceOfType (.prim (.bool b)) (.bool bty)
```